syncer/client.go

```diff
@@ -0,0 +1,277 @@
+			}
+
+			syncPeersLock.Lock()
+
+			syncPeers = append(syncPeers, status)
+
+			syncPeersLock.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	return syncPeers
+}
+
+func (m *syncPeerClient) GetPeerStatusUpdateCh() <-chan *NoForkPeer {
+	return m.peerStatusUpdateCh
+}
+
+func (m *syncPeerClient) GetPeerConnectionUpdateEventCh() <-chan *event.PeerEvent {
+	return m.peerConnectionUpdateCh
+}
+
+func (m *syncPeerClient) startGossip() error {
+	topic, err := m.network.NewTopic(statusTopicName, &proto.SyncPeerStatus{})
+	if err != nil {
+		return err
+	}
+
+	if err := topic.Subscribe(m.handleStatusUpdate); err != nil {
+		return fmt.Errorf("unable to subscribe to gossip topic, %w", err)
+	}
+
+	m.topic = topic
+
+	return nil
+}
+
+func (m *syncPeerClient) handleStatusUpdate(obj interface{}, from peer.ID) {
+	status, ok := obj.(*proto.SyncPeerStatus)
+	if !ok {
+		m.logger.Error("failed to cast gossiped message to txn")
+
+		return
+	}
+
+	if !m.network.IsConnected(from) {
+		if m.id != from.String() {
+			m.logger.Warn("received status from non-connected peer, ignore", "id", from)
+		}
+
+		return
+	}
+
+	m.peerStatusUpdateCh <- &NoForkPeer{
+		ID:       from,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(from),
+	}
+}
+
+func (m *syncPeerClient) startNewBlockProcess() {
+	m.subscription = m.blockchain.SubscribeEvents()
+
+	for event := range m.subscription.GetEventCh() {
+		if event.Type == blockchain.EventHead && len(event.NewChain) > 0 {
+			latest := m.blockchain.Header()
+			if latest == nil {
+				return
+			}
+
+			// Publish status
+			if err := m.topic.Publish(&proto.SyncPeerStatus{
+				Number: latest.Number,
+			}); err != nil {
+				m.logger.Warn("failed to publish status", "err", err)
+			}
+		}
+	}
+}
+
+func (m *syncPeerClient) startPeerEventProcess() {
+	peerEventCh, err := m.network.SubscribeCh()
+	if err != nil {
+		m.logger.Error("failed to subscribe", "err", err)
+
+		return
+	}
+
+	for e := range peerEventCh {
+		switch e.Type {
+		case event.PeerConnected, event.PeerDisconnected:
+			m.peerConnectionUpdateCh <- e
+		}
+	}
+}
+
+func (m *syncPeerClient) CloseStream(peerID peer.ID) error {
+	return m.network.CloseProtocolStream(SyncerProto, peerID)
+}
+
+func (m *syncPeerClient) GetBlocks(
+	ctx context.Context,
+	peerID peer.ID,
+	from uint64,
+) (<-chan *types.Block, error) {
+	clt, err := m.newSyncPeerClient(peerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create sync peer client: %w", err)
+	}
+
+	stream, err := clt.GetBlocks(ctx, &proto.GetBlocksRequest{
+		From: from,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to open GetBlocks stream: %w", err)
+	}
+
+	blockCh := make(chan *types.Block, 1)
+
+	go func() {
+		defer close(blockCh)
+
+		for {
+			protoBlock, err := stream.Recv()
+			if err != nil {
+				break
+			}
+
+			block, err := fromProto(protoBlock)
+			if err != nil {
+				m.logger.Warn("failed to decode a block from peer", "peerID", peerID, "err", err)
+
+				break
+			}
```

Comment:
What happens when the peer hangs in sending you blocks?

You should rely on a timeout here somewhere

Reply:
We can't set constant timeout here because it's impossible to expect how long it takes to finish sending all blocks. Then we need to start new goroutine and can reset timer if new block reached, otherwise you can call cancel context

Reply:
What do you think?

Reply:
That's exactly what I was thinking - have a base timeout for a block to reach you during bulk sync. What this timeout should be, is a different question 🙂 

@Aleksao998

Reply:
https://github.com/0xPolygon/polygon-edge/pull/591/commits/cb1ec52b9e95ad1eed924b8535b979cb73ff4bb8
I added

Reply:
Please don't resolve comments just by looking at the code. This fix introduced an issue where every few logs we are going to get and error log

Reply:
The issue was when we are doing watchSync, we are opening the stream and requesting all blocks that we do not have. Once we get all blocks we would get EOF response. All responses we are catching as err and we were logging this as error event though it is expected behavior
Fixed: 26daf07b2011c1cd3e2748531d2cae02238a0b7c